packages/dynamic_layouts/lib/src/staggered_layout.dart

```diff
@@ -0,0 +1,279 @@
+      _mainAxisCount[i] = 0;
+    }
+  }
+
+  /// The number of children in the cross axis.
+  final int crossAxisCount;
+
+  /// The number of logical pixels between each child along the main axis.
+  final double mainAxisSpacing;
+
+  /// The number of logical pixels between each child along the cross axis.
+  final double crossAxisSpacing;
+
+  /// The number of pixels from the leading edge of one tile to the trailing
+  /// edge of the same tile in the cross axis.
+  final double childCrossAxisExtent;
+
+  /// The axis along which the scroll view scrolls.
+  final Axis scrollDirection;
+
+  /// The collection of scroll offsets for every row or column across the main
+  /// axis. It includes the tiles' sizes and the spacing between them.
+  final Map<int, double> _scrollOffsetForMainAxis = <int, double>{};
+
+  /// The amount of tiles in every row or column across the main axis.
+  final Map<int, int> _mainAxisCount = <int, int>{};
+
+  /// Returns the row or column with the minimum extent for the next child to
+  /// be layed out.
+  int _getNextCrossAxisSlot() {
+    int nextCrossAxisSlot = 0;
+    double minScrollOffset = double.infinity;
+    _scrollOffsetForMainAxis.forEach((int mainAxisSeries, double scrollOffset) {
+      if (scrollOffset < minScrollOffset) {
+        nextCrossAxisSlot = mainAxisSeries;
+        minScrollOffset = scrollOffset;
+      }
+    });
+    return nextCrossAxisSlot;
+  }
+
+  /// Returns true if every child has crossed the [targetOffset] and false
+  /// if not.
+  @override
+  bool reachedTargetScrollOffset(double targetOffset) {
+    for (final double scrollOffset in _scrollOffsetForMainAxis.values) {
+      if (scrollOffset < targetOffset) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /// The infinite extent in the main axis lets
+  /// [RenderDynamicSliverGrid.performLayout()] know that the child will decide
+  /// its own size in that axis.
+  @override
+  DynamicSliverGridGeometry getGeometryForChildIndex(int index) {
+    return DynamicSliverGridGeometry(
+      scrollOffset: 0.0,
+      crossAxisOffset: 0.0,
+      mainAxisExtent: double.infinity,
+      crossAxisExtent: childCrossAxisExtent,
+    );
+  }
+
+  /// Returns an updated [DynamicSliverGridGeometry] with the actual size of
+  /// the child.
+  @override
+  DynamicSliverGridGeometry updateGeometryForChildIndex(
+      int index, Size childSize) {
+    final int crossAxisSlot = _getNextCrossAxisSlot();
+    final double currentScrollOffset = _scrollOffsetForMainAxis[crossAxisSlot]!;
+    final double childMainAxisExtent =
+        scrollDirection == Axis.vertical ? childSize.height : childSize.width;
+    _scrollOffsetForMainAxis[crossAxisSlot] =
+        childMainAxisExtent + _scrollOffsetForMainAxis[crossAxisSlot]!;
+
+    final DynamicSliverGridGeometry sliverGridGeometry =
+        DynamicSliverGridGeometry(
+      scrollOffset: currentScrollOffset +
+          _mainAxisCount[crossAxisSlot]! * mainAxisSpacing,
```

Comment:
Is `_mainAxisCount` meant to only facilitate the spacing?

Reply:
Yes, I figured it was necessary to keep track of the previous tiles for calculating the total offset from the accumulated spacing